aten/src/ATen/core/jit_type.h

```diff
@@ -391,6 +392,239 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
+  {
+  }
+
+  VaryingShape(const std::vector<int64_t>& vec)
+  : is_varying_size_(false)
+  , dims_(vec.begin(), vec.end())
+  {
+  }
+
+  VaryingShape(size_t size, bool is_varying_size = false)
+  : is_varying_size_(is_varying_size)
+  , dims_(size)
+  {
+  }
+
+  bool operator ==(const VaryingShape &other) const
+  {
+    if (is_varying_size_ != other.is_varying_size_)
+    {
+      return false;
+    }
+
+    return dims_ == dims_;
+  }
+
+  c10::optional<int64_t>& operator [] (int i)
+  {
+      if (is_varying_size_)
+      {
+        throw std::runtime_error("Shape isn't fixed");
+      }
+      return dims_[i];
+  }
+
+  const c10::optional<int64_t>& operator [] (int i) const
+  {
+      if (is_varying_size_)
+      {
+        throw std::runtime_error("Shape isn't fixed");
+      }
+      return dims_[i];
+  }
+
+  size_t size() const
+  {
+    if (is_varying_size())
+    {
+      throw std::runtime_error("Shape isn't fixed");
+    }
+    return dims_.size();
+  }
+
+  bool is_varying_size() const { return is_varying_size_; }
+
+
+  // we could probably rewrite this as a template specialization of merge
+  // but it will probably reduce the readibility.
+  VaryingShape merge(const VaryingShape& other) const;
+
+  std::vector<int64_t> freeze() const
+  {
+    std::vector<int64_t> frozen;
+    if (is_varying_size())
+    {
+      throw std::runtime_error("Shape isn't fixed");
+    }
+    for (auto d : dims_)
+    {
+      if (!d)
+      {
+        throw std::runtime_error("Shape isn't fixed");
+      }
+      frozen.push_back(d.value());
+    }
+
+    return frozen;
+  }
+
+private:
+  bool is_varying_size_;
+  std::vector<c10::optional<int64_t>> dims_;
+};
+
+using VaryingStrides = VaryingShape;
+
+struct ProfiledTensorType;
+using ProfiledTensorTypePtr = std::shared_ptr<ProfiledTensorType>;
+// This type represents a single Tensor with a specific size
+struct CAFFE2_API ProfiledTensorType : public TensorType {
+  template<typename ... T>
+  static ProfiledTensorTypePtr create( T&& ... all ) {
+    return ProfiledTensorTypePtr(new ProfiledTensorType( std::forward<T>(all)... )); // NOLINT(modernize-make-shared)
+  }
+
+  // overloaded create variadic template argument as it could not distinguish initializer list
+  static ProfiledTensorTypePtr create(at::ScalarType scalar_type, at::Device device, at::IntArrayRef sizes, at::IntArrayRef strides, bool requires_grad) {
+    return ProfiledTensorTypePtr(new ProfiledTensorType(scalar_type, device, sizes, strides, requires_grad)); // NOLINT(modernize-make-shared)
+  }
+
+  static ProfiledTensorTypePtr create(c10::optional<at::ScalarType> scalar_type, c10::optional<Device> device,const VaryingShape& sizes, const VaryingStrides& strides, c10::optional<bool> requires_grad)
+  {
+      return ProfiledTensorTypePtr(new ProfiledTensorType(scalar_type, device, sizes, strides, requires_grad)); // NOLINT(modernize-make-shared)
+  }
+
+
+  const VaryingShape& sizes() const { return sizes_; }
```

Comment:
We probably could. Unfortunately, we also need to worry about `strides_` which are similar to `sizes_` (they actually will probably need even more customization to be useful, which I think, we could add via subclassing `VaryingShape`) and I don't want to end up with `ProfiledTensorType`'s API that looks something like this: `getSizeDimensionAt`, `getStrideAt`. Also, it will be inconsistent with `CompleteTensorType` which `sizes()` returns a collection. How strongly do you feel about this?